sync_qubes-os.org_repo.sh

```diff
@@ -1,30 +1,62 @@
-#!/bin/sh
+#!/bin/bash --
+set -eu
+default_repos='current current-testing security-testing'
+usage () {
+    printf %s\\n "Usage: $0 RELEASE REPOS_TO_SYNC..." >&2
+    exit 1
+}
 
-pushd `dirname $0`
+dir=$(dirname "$0")
+pushd "$dir"
 
 #DRY="-n"
-[ "x$HOST" == "x" ] && HOST=yum.qubes-os.org
-[ "x$HOST_BASEDIR" == "x" ] && HOST_BASEDIR=/pub/qubes/repo/yum
-if [ -n "$1" ]; then
-    RELS_TO_SYNC=`basename "$1"`
+if [ "x${DRY+q}" = xq ] && [ "x$DRY" != x-n ]; then
+    echo 'Bad value for $DRY (must be -n or unset)'>&2
+    usage
+fi
+
+: "${DRY=}"
+
+[ "x${HOST-}" == "x" ] && HOST=yum.qubes-os.org
+if [[ ${#HOST} -gt 253 ]] || ! [[ "$HOST" =~ ^[A-Za-z0-9][A-Za-z0-9.-]+$ ]]; then
+    printf 'Invalid host %q\n' "$HOST">&2
+    usage
+fi
+[ "x${HOST_BASEDIR-}" == "x" ] && HOST_BASEDIR=/pub/qubes/repo/yum
+if ! [[ "$HOST_BASEDIR" =~ ^/[A-Za-z0-9/._-]+$ ]]; then
+    printf 'Invalid HOST_BASEDIR: %q\n' "$HOST_BASEDIR">&2
+    usage
+elif [ -n "${1-}" ]; then
+    RELS_TO_SYNC=$(basename "$1")
+    shift
 else
-    RELS_TO_SYNC="`readlink current-release|tr -d /`"
+    echo 'Release must be specified'>&2
+    usage
 fi
-REPOS_TO_SYNC="current current-testing security-testing"
+[[ "$RELS_TO_SYNC" =~ ^r[1-9]([.-][0-9a-z]*)?$ ]] || {
+    printf 'Bad release %q\n' "$RELS_TO_SYNC">&2
+    usage
+}
 
-for rel in $RELS_TO_SYNC; do
+for i; do
+    if ! [[ "$i" =~ ^[A-Za-z][A-Za-z0-9/._-]*$ ]]; then
+        echo "Invalid character in repository">&2
+        usage
+    fi
+done
 
-    rsync_args=
-    for repo in $REPOS_TO_SYNC; do
-        rsync_args="$rsync_args $rel/$repo"
-    done
-    echo "Syncing $rel..."
-    rsync $DRY --partial --progress --hard-links --exclude repodata -air $rsync_args $HOST:$HOST_BASEDIR/$rel/
-    rsync $DRY update_repo.sh update_repo-arg.sh mirrors.list $HOST:
-    for repo in $REPOS_TO_SYNC; do
-        [ -z "$DRY" ] && ssh $HOST sh -c "\"cd $HOST_BASEDIR; ~/update_repo-arg.sh $rel/$repo/dom0/fc* $rel/$repo/vm/*\""
-    done
+export PATH=/usr/bin:/bin
+rel=$RELS_TO_SYNC
 
+REPOS_TO_UPDATE=
+for i; do REPOS_TO_UPDATE+=" $i"; done
+echo "Syncing $rel..."
+for i; do
+    rsync $DRY --partial --progress --mkpath --hard-links -air -- "./$i" "$HOST:$HOST_BASEDIR/${i%/*}"
```

Comment:
`--mkpath` would be indeed useful here, but sadly it's too new to use. Any better idea than separate `ssh ... mkdir -p ...`?

Reply:
How new an rsync do we have?  Otherwise, I don’t have any better idea, although we only need a single call to `mkdir`.

Reply:
Fixed by using a single call to `mkdir`.

Reply:
Well, `--mkpath` is there since August this year, so not many distros will have it. Anyway, server has rsync 3.1.3.